File: 18.四数之和.java
import java.util.*;

/*
 * @lc app=leetcode.cn id=18 lang=java
 *
 * [18] 四数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        int n = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        if (n < 4)  return res;
        Arrays.sort(nums);
        int w, x, y, z;
        for (int i = 0; i < n - 3; ++i) {
            w = nums[i];
            if (i > 0 && w == nums[i - 1])  continue;
            for (int j = i + 1; j < n - 2; ++j) {
                x = nums[j];
                if (j > i + 1 && x == nums[j - 1])  continue;
                int l = j + 1, r = n - 1;
                while (l < r) {
                    y = nums[l];
                    z = nums[r];
                    int s = w + x + y + z;
                    if (s == target) {
                        res.add(new ArrayList<>(Arrays.asList(w, x, y, z)));
                        while (l < r && nums[l + 1] == nums[l]) ++l;
                        while (l < r && nums[r - 1] == nums[r]) --r;
                        ++l;
                        --r;
                    } else if (s > target) {
                        --r;
                    } else {
                        ++l;
                    }
                }
            }
        }
        return res;
    }
}
// @lc code=end

File: 18.四数之和.py
#
# @lc app=leetcode.cn id=18 lang=python3
#
# [18] 四数之和
#
# from typing import List

# @lc code=start
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        n = len(nums)
        if n < 4:
            return []
        nums.sort()
        res = []
        for i in range(n-3):
            w = nums[i]
            if i > 0 and w == nums[i-1]:
                continue    # avoid repetitions
            for j in range(i+1, n-2):
                x = nums[j]
                if j > i+1 and x == nums[j-1]:
                    continue    # avoid repetitions
                left, right = j+1, n-1
                while left < right:
                    y, z = nums[left], nums[right]
                    s = w + x + y + z
                    if s > target:
                        right -= 1
                    elif s < target:
                        left += 1
                    else:
                        res.append([w, x, y, z])
                        while left < right and nums[left] == nums[left+1]:
                            left += 1   # avoid repetitions
                        while left < right and nums[right] == nums[right-1]:
                            right -= 1  # avoid repetitions
                        left += 1
                        right -= 1
        return res

# @lc code=end
# s = Solution()
# print(s.fourSum([-1, 0, 1, 2, -1, -4], -1))
